File: Javascript/groupAnagrams.js
class Solution {
  groupAnagrams(arr) {
    let res = {};

    for (let str of arr) {
      let count = new Array(26).fill(0);

      for (let i = 0; i < str.length; i++) {
        count[str.charCodeAt(i) - 5] += 1;
      }
      if (res[count]) {
        res[count].push(str);
      } else {
        res[count] = [str];
      }
    }
    return Object.values(res);
  }
}

let sol = new Solution();
let arr = ["eat", "tea", "tan", "ate", "nat", "bat"];
let val = sol.groupAnagrams(arr);
console.log(val);

File: python/groupAnagrams.py
from collections import defaultdict
import collections


class Solution:
    def groupAnagrams(self, strs):
        # this adds lists as values
        # myDict = collections.defaultdict(tuple)
        myDict = {}

        for str in strs:
            count = [0] * 26
            for s in str:
                count[ord(s) - ord('a')] += 1
            if tuple(count) in myDict.keys():
                myDict[tuple(count)].append(str)
            else:
                myDict[tuple(count)] = [str]
        return myDict.values()


sol = Solution()
val = sol.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
print(val)


dict1 = collections.defaultdict(list)
dict1['name'].append('1')
dict1['name'].append('1')
print(dict1)
